filebeat/input/filestream/copytruncate_prospector.go

```diff
@@ -0,0 +1,344 @@
+}
+
+// addOriginalFile adds a new original file and its identifying information
+// to the bookkeeper.
+func (r rotatedFilestreams) addOriginalFile(path string, src loginp.Source) {
+	if _, ok := r.table[path]; ok {
+		return
+	}
+	r.table[path] = &rotatedFilestream{originalSrc: src, rotated: make([]rotatedFileInfo, 0, r.count)}
+}
+
+// isOriginalAdded checks if an original file has been found.
+func (r rotatedFilestreams) isOriginalAdded(path string) bool {
+	_, ok := r.table[path]
+	return ok
+}
+
+// originalSrc returns the original Source information of a given
+// original file path.
+func (r rotatedFilestreams) originalSrc(path string) loginp.Source {
+	return r.table[path].originalSrc
+}
+
+// addRotatedFile adds a new rotated file to the list and returns its index.
+// if a file is already added, the source is updated and the index is returned.
+func (r rotatedFilestreams) addRotatedFile(original, rotated string, src loginp.Source) int {
+	for idx, fi := range r.table[original].rotated {
+		if fi.path == rotated {
+			r.table[original].rotated[idx].src = src
+			return idx
+		}
+	}
+
+	r.table[original].rotated = append(r.table[original].rotated, rotatedFileInfo{rotated, src, time.Time{}})
+	r.sorter.sort(r.table[original].rotated)
+
+	for idx, fi := range r.table[original].rotated {
+		if fi.path == rotated {
+			return idx
+		}
+	}
+
+	return -1
+}
+
+// addRotatedFile adds a new rotated file to the list and returns its index.
+// if a file is already added, the source is updated and the index is returned.
+func (r rotatedFilestreams) removeRotatedFile(original, rotated string) {
+	for idx, fi := range r.table[original].rotated {
+		if fi.path == rotated {
+			r.table[original].rotated = append(r.table[original].rotated[:idx], r.table[original].rotated[idx+1:]...)
+			return
+		}
+	}
+}
+
+type copyTruncateFileProspector struct {
+	fileProspector
+	rotatedSuffix *regexp.Regexp
+	rotatedFiles  *rotatedFilestreams
+}
+
+// Run starts the fileProspector which accepts FS events from a file watcher.
+func (p *copyTruncateFileProspector) Run(ctx input.Context, s loginp.StateMetadataUpdater, hg loginp.HarvesterGroup) {
+	log := ctx.Logger.With("prospector", copyTruncateProspectorDebugKey)
+	log.Debug("Starting prospector")
+	defer log.Debug("Prospector has stopped")
+
+	defer p.stopHarvesterGroup(log, hg)
+
+	var tg unison.MultiErrGroup
+
+	tg.Go(func() error {
+		p.filewatcher.Run(ctx.Cancelation)
+		return nil
+	})
+
+	tg.Go(func() error {
+		ignoreInactiveSince := getIgnoreSince(p.ignoreInactiveSince, ctx.Agent)
+
+		for ctx.Cancelation.Err() == nil {
+			fe := p.filewatcher.Event()
+
+			if fe.Op == loginp.OpDone {
+				return nil
+			}
+
+			src := p.identifier.GetSource(fe)
+			switch fe.Op {
+			case loginp.OpCreate, loginp.OpWrite:
+				if fe.Op == loginp.OpCreate {
+					log.Debugf("A new file %s has been found", fe.NewPath)
+
+					err := s.UpdateMetadata(src, fileMeta{Source: fe.NewPath, IdentifierName: p.identifier.Name()})
```

Comment:
Do we need to update the state for a new file if `isFileIgnored` is going to filter out the file?